File: CS/VirtualSources.InfiniteAsyncSource/ViewModels/MainViewModel.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.ComponentModel;
using System.Linq;
using System.Runtime.CompilerServices;
using System.Threading.Tasks;
using DevExpress.Mvvm;
using DevExpress.Xpf.Data;

namespace VirtualSources.InfiniteAsyncSource.ViewModels {
    public class MainViewModel : ViewModelBase {
        private readonly ObservableCollection<Item> _items = new();
        public DevExpress.Xpf.Data.InfiniteAsyncSource Source { get => GetProperty(() => Source); set => SetProperty(() => Source, value); }

        public MainViewModel() {
            _items = new ObservableCollection<Item>(Enumerable.Range(0, 10).Select(i =>
                new Item { Id = i, Name = $"Item {i}", CreatedAt = DateTime.Now.AddDays(i) }));

            Source = new DevExpress.Xpf.Data.InfiniteAsyncSource();
            Source.CustomProperties =
                new PropertyDescriptorCollection(TypeDescriptor.GetProperties(typeof(Item)).Cast<PropertyDescriptor>()
                                                               .ToArray());

            Source.FetchRows += (_, e) => {
                if (_items != null)
                    e.Result = Task.FromResult(new FetchRowsResult(_items.Cast<object>().ToArray(), false));
            };
        }
    }

    public class Item : BindableBase {
        private readonly Dictionary<string, object> _customFieldValues = new();
        public DateTime CreatedAt { get => GetProperty(() => CreatedAt); set => SetProperty(() => CreatedAt, value); }
        public int Id { get => GetProperty(() => Id); set => SetProperty(() => Id, value); }
        public string Name { get => GetProperty(() => Name); set => SetProperty(() => Name, value); }

        public object this[string fieldName] {
            get {
                _customFieldValues.TryGetValue(fieldName, out var value);
                return value;
            }

            set {
                _customFieldValues[fieldName] = value;
                RaisePropertyChanged(fieldName);
            }
        }
    }
}

File: VB/VirtualSources.InfiniteAsyncSource/ViewModels/MainViewModel.vb
Imports System.Collections.Generic
Imports System.Collections.ObjectModel
Imports System.ComponentModel
Imports System.Linq
Imports System.Runtime.CompilerServices
Imports System.Threading.Tasks
Imports DevExpress.Mvvm.CodeGenerators
Imports DevExpress.Xpf.Data

Namespace VirtualSources.InfiniteAsyncSource.ViewModels

    <GenerateViewModel>
    Public Partial Class MainViewModel

        Private ReadOnly _items As ObservableCollection(Of Item) = New(_, _)()

        <GenerateProperty>
        Private _source As DevExpress.Xpf.Data.InfiniteAsyncSource

        Public Sub New()
            _items = New ObservableCollection(Of Item)(Enumerable.Range(0, 10).[Select](Function(i) New Item With {.Id = i, .Name = $"Item {i}", .CreatedAt = Date.Now.AddDays(i)}))
            Source = New DevExpress.Xpf.Data.InfiniteAsyncSource()
            Source.CustomProperties = New PropertyDescriptorCollection(TypeDescriptor.GetProperties(GetType(Item)).Cast(Of PropertyDescriptor)().ToArray())
            Source.FetchRows += Function(__, e)
                If _items IsNot Nothing Then e.Result = Task.FromResult(New FetchRowsResult(_items.Cast(Of Object)().ToArray(), False))
            End Function
        End Sub
    End Class

    <GenerateViewModel>
    Public Partial Class Item
        Implements INotifyPropertyChanged

        Private ReadOnly _customFieldValues As Dictionary(Of String, Object) = New(_, _)()

        <GenerateProperty>
        Private _createdAt As Date

        <GenerateProperty>
        Private _id As Integer

        <GenerateProperty>
        Private _name As String

        Default Public Property Item(ByVal fieldName As String) As Object
            Get
                Dim value As Object = Nothing
                _customFieldValues.TryGetValue(fieldName, value)
                Return value
            End Get

            Set(ByVal value As Object)
                _customFieldValues(fieldName) = value
                OnPropertyChanged(fieldName)
            End Set
        End Property

        Public Event PropertyChanged As PropertyChangedEventHandler Implements INotifyPropertyChanged.PropertyChanged

        Protected Overridable Sub OnPropertyChanged(<CallerMemberName> ByVal Optional propertyName As String = Nothing)
            RaiseEvent PropertyChanged(Me, New PropertyChangedEventArgs(propertyName))
        End Sub
    End Class
End Namespace
